ReadMe
                       SQL-Routine-SQLParser
----------------------------------------------------------------------

by Darren Duncan <perl@DarrenDuncan.net>

CONTENTS

* Preface
* Keeping Up To Date
* Dependencies
* Support
* License And Copyright
* Acknowledgements

PREFACE

This distribution features the Perl 5 module named
"SQL::Routine::SQLParser"; see the file lib/SQL/Routine/SQLParser.pm for
the main documentation, which is in POD format.  The distribution also
includes a ChangeLog file to say what changes already happened, and a TODO
file to say what changes have yet to happen.  The file INSTALL should help
with installation issues.  Read below for how to get support or keep up to
date, know this distribution's dependencies, or read the credits or
copyright and license.

KEEPING UP TO DATE

My module set is constantly under development.  The canonical copies are
all linked to through my website, "http://www.DarrenDuncan.net/", on the
page called "Perl Libraries I Made" (name subject to change).

For those of you that are interested in the bleeding edge, I now have a
public Subversion repository at "http://svn.utsl.gen.nz/trunk/Rosetta/";
this repository contains all of my newer modules (whose problem domain is
databases), but not my older modules (which are not about databases).

You can download slightly older but more stable versions of all of my
modules in standard tarred and gzipped distributions, either from the
public CPAN archives, "http://www.cpan.org/authors/id/D/DU/DUNCAND/", or
from my own web server, "http://www.DarrenDuncan.net/d/perl/".

CPAN also extracts the POD from modules and maintains that documentation in
an organized and searchable fashion.  Their main search site is
"http://search.cpan.org/", and my own DUNCAND author page is
"http://search.cpan.org/~duncand/"; distributions are also there.

You can download my oldest distributions from BACKPAN, at
"http://history.perl.org/backpan/authors/id/D/DU/DUNCAND/", or from my web
server, at "http://www.DarrenDuncan.net/d/perl/archives/".

DEPENDENCIES

This distribution requires any version of Perl 5.x.y that is at least
5.8.1.

It also requires the Perl modules 'version' and 'only', which would
conceptually be built-in to Perl, but aren't, so they are on CPAN instead.

It also requires these modules that are on CPAN: Locale::KeyedText '1.6.0-'
(for error messages), SQL::Routine '0.70.0-'.

This distribution also requires Test::More for its test suite, and that
module is bundled with Perl.

SUPPORT

Currently I don't have any support arranged with other people, lists,
newsgroups, or otherwise.  Feel free to ask me if you can't figure things
out on your own, or another person whom you know has used this.  I may
start a mailing list for support issues later, so that users of my modules
can help each other with them.

LICENCE AND COPYRIGHT

This file is part of the SQL::Routine::SQLParser reference implementation
of a SQL:2003 string parser that uses the SQL::Routine database portability
library.

SQL::Routine::SQLParser is Copyright (c) 2002-2005, Darren R. Duncan.  All
rights reserved.  Address comments, suggestions, and bug reports to
perl@DarrenDuncan.net, or visit http://www.DarrenDuncan.net/ for more
information.

SQL::Routine::SQLParser is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License (GPL) as
published by the Free Software Foundation (http://www.fsf.org/); either
version 2 of the License, or (at your option) any later version.  You
should have received a copy of the GPL as part of the
SQL::Routine::SQLParser distribution, in the file named "GPL"; if not,
write to the Free Software Foundation, Inc., 51 Franklin St, Fifth Floor,
Boston, MA  02110-1301, USA.

Linking SQL::Routine::SQLParser statically or dynamically with other
modules is making a combined work based on SQL::Routine::SQLParser.  Thus,
the terms and conditions of the GPL cover the whole combination.  As a
special exception, the copyright holders of SQL::Routine::SQLParser give
you permission to link SQL::Routine::SQLParser with independent modules,
regardless of the license terms of these independent modules, and to copy
and distribute the resulting combined work under terms of your choice,
provided that every copy of the combined work is accompanied by a complete
copy of the source code of SQL::Routine::SQLParser (the version of
SQL::Routine::SQLParser used to produce the combined work), being
distributed under the terms of the GPL plus this exception.  An independent
module is a module which is not derived from or based on
SQL::Routine::SQLParser, and which is fully useable when not linked to
SQL::Routine::SQLParser in any form.

Any versions of SQL::Routine::SQLParser that you modify and distribute must
carry prominent notices stating that you changed the files and the date of
any changes, in addition to preserving this original copyright notice and
other credits.  SQL::Routine::SQLParser is distributed in the hope that it
will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty
of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

While it is by no means required, the copyright holders of
SQL::Routine::SQLParser would appreciate being informed any time you create
a modified version of SQL::Routine::SQLParser that you are willing to
distribute, because that is a practical way of suggesting improvements to
the standard version.

ACKNOWLEDGEMENTS

None besides myself so far.

FIN

Share and Enjoy!


Changes
                       SQL-Routine-SQLParser
----------------------------------------------------------------------

2005-09-28   Darren Duncan <perl@DarrenDuncan.net>

    * SQL-Routine-SQLParser Release 0.2.2 (SVN r878).

    * New code file versions are: SQLParser.pm 0.2.2 and en.pm 0.2.2.

    * From now on, any ordinary 'use [|only ]Foo' statements that appear
    within *.pm files will be located in different places than before; the
    'use' statements for modules that do not export anything, particularly
    object oriented modules, will appear near the top of the file, above
    all package declarations, but just below the use-pragma statements; the
    'use' statements for modules whose exported functions we are using will
    be placed just below the declarations of each package in which the
    relevant exported functions are used.

    * Removed any line-trailing whitespace from all distribution files.

    * Reformatted all code by swapping various string quoting delimiters.

    * Updated SQLParser.pm and t_SRT_SP_Util.pm to replace all of their
    'unless' conditionals with equivalent 'if' conditionals.

    * Updated SQLParser.pm to rename all of its 'foreach' loops to 'for'.

    * Reformatted all code to change any "if( ... ) {" to "if (...) {".

    * Added named constant $EMPTY_STR to SQLParser.pm.

    * Reformatted all code so that any uses of the named unary operators
    [defined|ref] no longer have parenthesis around their argument.

    * Updated SQLParser.pm to replace 1 "<condition> and return ...;" with
    "return ... if <condition>;".

    * Updated t_SRT_SP_Util.pm to replace 1 "if ... return ..." with its
    postfix-if equivalent.

    * Updated SQLParser.pm to rearrange 3 conditional '_throw...' statements
    so they have postfix-if conditionals.

    * Reformatted all code, in SQLParser.pm, so that any uses of the
    built-in operators or functions named [bless|join] no longer have
    parenthesis around their argument lists; either the parens were just
    removed, or they were moved to surround both the operator/function name
    and its arguments.

    * Split up some long lines in en.pm.

    * Updated any DEPENDENCIES documentation to reformat any indented
    sections as paragraphs, and show version ranges like 'only' takes them.

    * Reformatted all code so that every occurance of the string
    concatenation operator (.) has a space between the operator and each of
    its 2 arguments, rather than their all being in contact.  Also
    reformatted any lines that are split on this operator so the operator
    appears at the start of the second line, rather than the end of the
    first line.

    * Other miscellaneous code line alignments and splitting.

2005-09-12   Darren Duncan <perl@DarrenDuncan.net>

    * SQL-Routine-SQLParser Release 0.2.1 (SVN r808).

    * New code file versions are: SQLParser.pm 0.2.1 and en.pm 0.2.1.

    * Updated all POD-containing files to re-wrap any non-indented POD
    paragraphs to a 75 character width, which is 5 less than the 80
    character width they were wrapped to before.  This change should make
    it easier to copy and paste a diff or patch of this documentation into
    an email message, where it is quoted at least once, without any line
    wrapping occurring.  Likewise, these standard documentation files were
    re-wrapped to 75 characters: ReadMe, INSTALL, LGPL|GPL, Changes, TODO.

    * Revised this Changes file to replace all detail entries for releases
    0.01 thru 0.2.0 with a significant release list for the same period;
    also added references to when all 'Rosetta developer release' (#s 1-3)
    were.

2005-09-08   Darren Duncan <perl@DarrenDuncan.net>

    * SQL-Routine-SQLParser Release 0.2.0 (SVN r787), containing
    SQL::Routine::SQLParser 0.2.0, was released on CPAN.  This is the first
    release of this distribution and its modules for which they had 3-part
    version numbers, rather than floating point version numbers.

    * These were the current versions cited by the public announcement for
    Rosetta/SQL-Routine developer release #3.

2005-06-25   Darren Duncan <perl@DarrenDuncan.net>

    * SQL-Routine-SQLParser Release 0.01, containing
    SQL::Routine::SQLParser 0.01, was released on CPAN; this is the first
    release of the SQL-Routine-SQLParser distribution.  This is the first
    release of any distribution to contain SQL::Routine::SQLParser.  This
    is the first release of any distribution for which some of the existing
    work that became the SQL-Routine-SQLParser distribution was in the form
    of executable code, and not just design documentation.  This is the
    only release of this distribution and its modules for which they had
    floating point version numbers, rather than 3-part version numbers.

2005-04-03   Darren Duncan <perl@DarrenDuncan.net>

    * These did not exist and were not cited by the public announcement for
    SQL-Routine/Rosetta developer release #2.

2004-08-06   Darren Duncan <perl@DarrenDuncan.net>

    * These did not exist and were not cited by the public announcement for
    Rosetta/SQL-SyntaxModel developer release #1.


t/lib/t_SRT_SP_Util.pm
#!perl
use 5.008001; use utf8; use strict; use warnings;

# This module is used when testing SQL::Routine::SQLParser.
# It contains some utility methods used by the various SRT_SP_*.t scripts.

package # hide this class name from PAUSE indexer
t_SRT_SP_Util;

######################################################################

sub message {
    my (undef, $detail) = @_;
    print "# $detail\n";
}

######################################################################

sub error_to_string {
    my (undef, $message) = @_;
    if (ref $message and UNIVERSAL::isa( $message, 'Locale::KeyedText::Message' )) {
        my $translator = Locale::KeyedText->new_translator(
            ['SQL::Routine::SQLParser::L::', 'SQL::Routine::L::'], ['en'] );
        my $user_text = $translator->translate_message( $message );
        return q{internal error: can't find user text for a message: }
            . $message->as_string() . ' ' . $translator->as_string()
            if !$user_text;
        return $user_text;
    }
    return $message; # if this isn't the right kind of object
}

######################################################################

1;


lib/SQL/Routine/SQLParser/L/en.pm
#!perl
use 5.008001; use utf8; use strict; use warnings;

package SQL::Routine::SQLParser::L::en;
use version; our $VERSION = qv('0.2.2');

######################################################################

my $CSP = 'SQL::Routine::SQLParser';

my %text_strings = (
    'SRT_SP_METH_ARG_UNDEF' =>
        $CSP . q[.{METH}(): ]
        . q[undefined (or missing) {ARGNM} argument],
    'SRT_SP_METH_ARG_NO_ARY' =>
        $CSP . q[.{METH}(): ]
        . q[invalid {ARGNM} argument; ]
        . q[it is not a Array ref, but rather is "{ARGVL}"],
    'SRT_SP_METH_ARG_NO_HASH' =>
        $CSP . q[.{METH}(): ]
        . q[invalid {ARGNM} argument; ]
        . q[it is not a Hash ref, but rather is "{ARGVL}"],
    'SRT_SP_METH_ARG_NO_NODE' =>
        $CSP . q[.{METH}(): ]
        . q[invalid {ARGNM} argument; ]
        . q[it is not a Node object, but rather is "{ARGVL}"],

    'SRT_SP_METH_ARG_WRONG_NODE_TYPE' =>
        $CSP . q[.{METH}(): ]
        . q[invalid {ARGNM} argument; ]
        . q[it is not a "{EXPNTYPE}" Node, but rather is a "{ARGNTYPE}" Node],
);

######################################################################

sub get_text_by_key {
    my (undef, $msg_key) = @_;
    return $text_strings{$msg_key};
}

######################################################################

1;
__END__

=encoding utf8

=head1 NAME

SQL::Routine::SQLParser::L::en - Localization of SQL::Routine::SQLParser for English

=head1 VERSION

This document describes SQL::Routine::SQLParser::L::en version 0.2.2.

=head1 SYNOPSIS

    use Locale::KeyedText;
    use SQL::Routine::SQLParser;

    # do work ...

    my $translator = Locale::KeyedText->new_translator(
        ['SQL::Routine::SQLParser::L::', 'SQL::Routine::L::'], ['en'] );

    # do work ...

    eval {
        # do work with SQL::Routine::SQLParser, which may throw an exception ...
    };
    if (my $error_message_object = $@) {
        # examine object here if you want and programmatically recover...

        # or otherwise do the next few lines...
        my $error_user_text = $translator->translate_message( $error_message_object );
        # display $error_user_text to user by some appropriate means
    }

    # continue working, which may involve using SQL::Routine::SQLParser some more ...

=head1 DESCRIPTION

The SQL::Routine::SQLParser::L::en Perl 5 module contains localization data
for SQL::Routine::SQLParser.  It is designed to be interpreted by
Locale::KeyedText.  Besides localizing generic error messages that
SQL::Routine::SQLParser produces itself, this file also provides a
ready-made set of generic database error strings that can be thrown by any
SQL::Routine::SQLParser Engine.

This class is optional and you can still use SQL::Routine::SQLParser
effectively without it, especially if you plan to either show users
different error messages than this class defines, or not show them anything
because you are "handling it".

=head1 FUNCTIONS

=head2 get_text_by_key( MSG_KEY )

    my $user_text_template = SQL::Routine::SQLParser::L::en->get_text_by_key( 'foo' );

This function takes a Message Key string in MSG_KEY and returns the
associated user text template string, if there is one, or undef if not.

=head1 DEPENDENCIES

This module requires any version of Perl 5.x.y that is at least 5.8.1.

It also requires the Perl module L<version>, which would conceptually be
built-in to Perl, but isn't, so it is on CPAN instead.

This module has no enforced dependencies on L<Locale::KeyedText>, which is
on CPAN, or on L<SQL::Routine::SQLParser>, which is in the current
distribution, but it is designed to be used in conjunction with them.

=head1 INCOMPATIBILITIES

None reported.

=head1 SEE ALSO

L<perl(1)>, L<Locale::KeyedText>, L<SQL::Routine::SQLParser>.

=head1 BUGS AND LIMITATIONS

The structure of this module is trivially simple and has no known bugs.

However, the locale data that this module contains may be subject to large
changes in the future; you can determine the likeliness of this by
examining the development status and/or BUGS AND LIMITATIONS documentation
of the other module that this one is localizing; there tends to be a high
correlation in the rate of change between that module and this one.

=head1 AUTHOR

Darren R. Duncan (C<perl@DarrenDuncan.net>)

=head1 LICENCE AND COPYRIGHT

This file is part of the SQL::Routine::SQLParser reference implementation
of a SQL:2003 string parser that uses the SQL::Routine database portability
library.

SQL::Routine::SQLParser is Copyright (c) 2002-2005, Darren R. Duncan.  All
rights reserved.  Address comments, suggestions, and bug reports to
C<perl@DarrenDuncan.net>, or visit L<http://www.DarrenDuncan.net/> for more
information.

SQL::Routine::SQLParser is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License (GPL) as
published by the Free Software Foundation (L<http://www.fsf.org/>); either
version 2 of the License, or (at your option) any later version.  You
should have received a copy of the GPL as part of the
SQL::Routine::SQLParser distribution, in the file named "GPL"; if not,
write to the Free Software Foundation, Inc., 51 Franklin St, Fifth Floor,
Boston, MA  02110-1301, USA.

Linking SQL::Routine::SQLParser statically or dynamically with other
modules is making a combined work based on SQL::Routine::SQLParser.  Thus,
the terms and conditions of the GPL cover the whole combination.  As a
special exception, the copyright holders of SQL::Routine::SQLParser give
you permission to link SQL::Routine::SQLParser with independent modules,
regardless of the license terms of these independent modules, and to copy
and distribute the resulting combined work under terms of your choice,
provided that every copy of the combined work is accompanied by a complete
copy of the source code of SQL::Routine::SQLParser (the version of
SQL::Routine::SQLParser used to produce the combined work), being
distributed under the terms of the GPL plus this exception.  An independent
module is a module which is not derived from or based on
SQL::Routine::SQLParser, and which is fully useable when not linked to
SQL::Routine::SQLParser in any form.

Any versions of SQL::Routine::SQLParser that you modify and distribute must
carry prominent notices stating that you changed the files and the date of
any changes, in addition to preserving this original copyright notice and
other credits.  SQL::Routine::SQLParser is distributed in the hope that it
will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty
of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

While it is by no means required, the copyright holders of
SQL::Routine::SQLParser would appreciate being informed any time you create
a modified version of SQL::Routine::SQLParser that you are willing to
distribute, because that is a practical way of suggesting improvements to
the standard version.

=cut


lib/SQL/Routine/SQLParser.pm
#!perl
use 5.008001; use utf8; use strict; use warnings;

use only 'Locale::KeyedText' => '1.6.0-';
use only 'SQL::Routine' => '0.70.0-';

package SQL::Routine::SQLParser;
use version; our $VERSION = qv('0.2.2');

######################################################################
######################################################################

# These are constant values used by this module.
my $EMPTY_STR = q{};

######################################################################

sub new {
    my ($class) = @_;
    my $parser = bless {}, ref $class || $class;



    return $parser;
}

######################################################################
# This is a 'protected' method; only sub-classes should invoke it.

sub _throw_error_message {
    my ($parser, $msg_key, $msg_vars) = @_;
    # Throws an exception consisting of an object.
    ref $msg_vars eq 'HASH' or $msg_vars = {};
    for my $var_key (keys %{$msg_vars}) {
        if (ref $msg_vars->{$var_key} eq 'ARRAY') {
            $msg_vars->{$var_key} = 'PERL_ARRAY:[' . (join q{,},map {$_||$EMPTY_STR} @{$msg_vars->{$var_key}}) . ']';
        }
    }
    die Locale::KeyedText->new_message( $msg_key, $msg_vars );
}

sub _assert_arg_node_type {
    my ($parser, $meth_name, $arg_name, $exp_node_types, $arg_value) = @_;
    $parser->_throw_error_message( 'SRT_SP_METH_ARG_UNDEF',
        { 'METH' => $meth_name, 'ARGNM' => $arg_name } )
        if !defined $arg_value;
    $parser->_throw_error_message( 'SRT_SP_METH_ARG_NO_NODE',
        { 'METH' => $meth_name, 'ARGNM' => $arg_name, 'ARGVL' => $arg_value } )
        if !ref $arg_value or !UNIVERSAL::isa( $arg_value, 'SQL::Routine::Node' );
    return
        if @{$exp_node_types} == 0; # any Node type is acceptable
    my $given_node_type = $arg_value->get_node_type();
    $parser->_throw_error_message( 'SRT_SP_METH_ARG_WRONG_NODE_TYPE',
        { 'METH' => $meth_name, 'ARGNM' => $arg_name,
        'EXPNTYPE' => $exp_node_types, 'ARGNTYPE' => $given_node_type } )
        if !grep { $given_node_type eq $_ } @{$exp_node_types};
    # If we get here, $arg_value is acceptable to the method.
}

######################################################################
######################################################################

1;
__END__

=encoding utf8

=head1 NAME

SQL::Routine::SQLParser - Parse ANSI/ISO SQL:2003 and other SQL variants

=head1 VERSION

This document describes SQL::Routine::SQLParser version 0.2.2.

=head1 SYNOPSIS

I<This documentation section will be written later.>

=head1 DESCRIPTION

This module is a reference implementation of fundamental SQL::Routine
features.

The SQL::Routine::SQLParser Perl 5 module is a functional but quickly built
SQL::Routine utility class that converts one or more SQL strings into a set
of related SQL::Routine Nodes that are functionally equivalent, and can be
used for a variety of tasks that SQL would normally be used with, but
saving the users from having to parse the string SQL themselves.  This
class will parse multiple string SQL variants; usually it can determine the
intent of a SQL string based soley from examining the string itself, but in
cases where the string is ambiguous as to what SQL variant it is, this
class takes arguments that let you give it hints to resolve the ambiguity;
by default it will resolve ambiguities by whatever decision results in
something closer to the ANSI/ISO SQL:2003 (or 1999 or 1992) standard.

SQL::Routine::SQLParser is designed to implement common functionality for
multiple Rosetta Engine classes (such as Rosetta::Engine::Generic) allowing
them to focus more on the non-SQL specific aspects of their work.  A
Rosetta Engine would typically invoke this class when reverse-engineering a
database schema where the database returns the schema object descriptors as
SQL strings ('create' statements usually) rather than the information being
in an "information schema".  This class can also be used by code on the
application-side of a Rosetta::Interface tree (such as
Rosetta::Emulator::DBI); for example, a module that emulates an older
database interface which wants to accept database commands as SQL strings
can use this module to parse those. (For your reference, see also the
SQL::Routine::SQLBuilder module, which implements the inverse functionality
to SQLParser, and is used in both of the same places.)

I<CAVEAT:  SIGNIFICANT PORTIONS OF THIS MODULE ARE NOT WRITTEN YET.>

=head1 CONSTRUCTOR FUNCTIONS

This function is stateless and can be invoked off of either this module's
name or an existing module object, with the same result.

=head2 new()

    my $parser = SQL::Routine::SQLParser->new();
    my $parser2 = $parser->new();

This "getter" function/method will create and return a single
SQL::Routine::SQLParser (or subclass) object.  All of this object's
properties are set to default values that should cause the object to
resolve SQL parsing ambiguities in a manner leaning towards the SQL:2003
standard.

=head1 DEPENDENCIES

This module requires any version of Perl 5.x.y that is at least 5.8.1.

It also requires the Perl modules L<version> and L<only>, which would
conceptually be built-in to Perl, but aren't, so they are on CPAN instead.

It also requires these modules that are on CPAN: L<Locale::KeyedText>
'1.6.0-' (for error messages), L<SQL::Routine> '0.70.0-'.

=head1 INCOMPATIBILITIES

None reported.

=head1 SEE ALSO

L<perl(1)>, L<SQL::Routine::SQLParser::L::en>, L<Locale::KeyedText>,
L<SQL::Routine>, L<SQL::Routine::SQLBuilder>, L<Rosetta::Engine::Generic>,
L<Rosetta::Emulator::DBI>, L<DBIx::MyParse>.

=head1 BUGS AND LIMITATIONS

This module is currently in pre-alpha development status, meaning that some
parts of it will be changed in the near future, perhaps in incompatible
ways.

=head1 AUTHOR

Darren R. Duncan (C<perl@DarrenDuncan.net>)

=head1 LICENCE AND COPYRIGHT

This file is part of the SQL::Routine::SQLParser reference implementation
of a SQL:2003 string parser that uses the SQL::Routine database portability
library.

SQL::Routine::SQLParser is Copyright (c) 2002-2005, Darren R. Duncan.  All
rights reserved.  Address comments, suggestions, and bug reports to
C<perl@DarrenDuncan.net>, or visit L<http://www.DarrenDuncan.net/> for more
information.

SQL::Routine::SQLParser is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License (GPL) as
published by the Free Software Foundation (L<http://www.fsf.org/>); either
version 2 of the License, or (at your option) any later version.  You
should have received a copy of the GPL as part of the
SQL::Routine::SQLParser distribution, in the file named "GPL"; if not,
write to the Free Software Foundation, Inc., 51 Franklin St, Fifth Floor,
Boston, MA  02110-1301, USA.

Linking SQL::Routine::SQLParser statically or dynamically with other
modules is making a combined work based on SQL::Routine::SQLParser.  Thus,
the terms and conditions of the GPL cover the whole combination.  As a
special exception, the copyright holders of SQL::Routine::SQLParser give
you permission to link SQL::Routine::SQLParser with independent modules,
regardless of the license terms of these independent modules, and to copy
and distribute the resulting combined work under terms of your choice,
provided that every copy of the combined work is accompanied by a complete
copy of the source code of SQL::Routine::SQLParser (the version of
SQL::Routine::SQLParser used to produce the combined work), being
distributed under the terms of the GPL plus this exception.  An independent
module is a module which is not derived from or based on
SQL::Routine::SQLParser, and which is fully useable when not linked to
SQL::Routine::SQLParser in any form.

Any versions of SQL::Routine::SQLParser that you modify and distribute must
carry prominent notices stating that you changed the files and the date of
any changes, in addition to preserving this original copyright notice and
other credits.  SQL::Routine::SQLParser is distributed in the hope that it
will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty
of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

While it is by no means required, the copyright holders of
SQL::Routine::SQLParser would appreciate being informed any time you create
a modified version of SQL::Routine::SQLParser that you are willing to
distribute, because that is a practical way of suggesting improvements to
the standard version.

=cut
